packages/core/src/node/dependency-download.ts

```diff
@@ -0,0 +1,60 @@
+// *****************************************************************************
+// Copyright (C) 2023 TypeFox and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { RequestOptions } from 'https';
+import { MaybePromise, URI } from 'src/common';
+
+/**
+ * contribution used for downloading prebuild nativ dependency when connecting to a remote machine with a different system
+ */
+export interface DependencyDownloadContribution {
+    httpOptions?: RequestOptions;
+    skipUnzip?: boolean
+
+    getDownloadUrl(remoteOS: string, theiaVersion: string | undefined): MaybePromise<string>;
```

Comment:
Suggestion: Not every native dependency will be placed into the same directory. For example, the `spawn-helper` from `node-pty` will go into `lib/build/Release`, whereas `pty.node` will be placed in `lib/backend/native`.

Therefore, we need to include more instructions on how to place the downloaded dependencies in the resulting build. Something like:

```ts
export interface FileDependencyResult {
    targetFile: string;
    /** Some files needs to be made executable on the target system */
    chmod?: number;
}

export interface FileDependencyDownload {
    /** Defaults to `true` */
    unzip?: boolean;
    file: FileDependencyResult
    downloadHandler: string | Buffer | () => Promise<Buffer | string>
}

// Directories are expected to be in a zipped format anyway
// We always unzip them and call `files` on each contained file
export interface DirectoryDependencyDownload {
    files: (path: string) => FileDependencyResult;
    downloadHandler: string | Buffer | () => Promise<Buffer | string>
}
```

Also, we can replace `httpOptions` and the arguments for this function with something better:

```ts
export interface DownloadOptions {
    remoteOS: string;
    theiaVersion: string;
    /** These are the `NodeRequestOptions` for the `NodeRequestService` */
    download: (params: string | NodeRequestOptions) => Promise<Buffer>
}
```
```suggestion
    getDownloadUrl(options: DownloadOptions): MaybePromise<FileDependencyDownload | DirectoryDependencyDownload>;
```

Reply:
i meant this feature for just downloading the dependencies. Placing the dependencies correctly is probably something the actual copying step should do. So the copying step would get the folder where the dependencies are and could then put them into the correct location

Reply:
I agree. However, the SSH copy mechanism still needs information on where to put these files. I would like to see a pipeline, in which this information doesn't need to be duplicated. I.e. there's only a single contribution for each native dependency and all other systems (like SSH copy or WSL copy, etc.) can just reuse the information from the contributions. We should use the aforementioned `FileDependencyResult` for this.